templates/README.md

```diff
@@ -0,0 +1,260 @@
+    flavor.template.cnv.io/tiny: "false"
+
+    # CNV Template type to separate the use cases for base OS,
+    # flavor, sizing templates and templates created from
+    # running or imported VMs.
+    # The supported values are currently: base and config
+    template.cnv.io/type: "base"
+
+# Parameters must come from a subset of well known names
+# so the UI can properly work with those.
+parameters:
+- name: NAME
+  description: VM name
+  generate: expression
+  from: 'vm-[A-Za-z0-9]{8}'
+- name: MEMORY_SIZE
+  description: Memory size
+  value: 1Gi
+
+objects:
+# The full VM template with placeholders for either scalars like memory
+# or yaml structures like disks
+- apiVersion: kubevirt.io/v1alpha2
+  kind: VirtualMachine
+  metadata:
+    name: ${NAME}
+  spec:
+    running: false
+    template:
+      spec:
+        domain:
+          resources:
+            initial:
+              memory: ${MEMORY_SIZE}Mi
+          devices:
+            disks:
+            # This should be interpreted as a template disk by the UI,
+            # thanks to the template.cnv.io/default annotations
+            # This must still result in a bootable VM when used as is.
+            # This way we can both use Templates for creating new VM as well
+            # as for converting an existing VM to a template
+            - name: default-disk
+              disk:
+                dev: vda
+              volumeName: default-volume
+
+            interfaces:
+            # This should be interpreted as a template network by the UI,
+            # thanks to the template.cnv.io/default annotations
+            # This must still result in a bootable VM when used as is.
+            # This way we can both use Templates for creating new VM as well
+            # as for converting an existing VM to a template
+            - type: pod-network
+              name: default-nic
+              model:
+                type: virtio
+
+        networks:
+          - name: default-network
+            resource:
+              resourceName: bridge.network.kubevirt.io/cnvmgmt
+
+        volumes:
+            # This should be interpreted as a template volume by the UI,
+            # thanks to the template.cnv.io/default annotations
+            # This must still result in a bootable VM when used as is.
+            # This way we can both use Templates for creating new VM as well
+            # as for converting an existing VM to a template
+          - name: default-volume
+            registryDisk:
+              image: test/image
+
+```
+
+### VM.yaml
+
+The above template should result in the following VM object:
+
+```YAML
+# vim: sts=2 sw=2 et
+apiVersion: kubevirt.io/v1alpha2
+kind: VirtualMachine
+metadata:
+  name: windows-10-1
+  annotations:
+    # Data used for the UI, ignored by kubevirt
+    # Arbitrary format as needed to be able to
+    # repopulate the UI or the template processor
+    # and get the same output
+    template.cnv.io/original:
```

Comment:
Actually, I need to change this section, because it won't work this way. The value has to be a string. I will have to change it to something like parameter.template.cnv.io/{param-name}: value

Reply:
I'd actually use a single annotation but then serialize it in json.This will be more predictive, as we know we only have one annotation.